### PR TITLE
Odr warning overhaul

### PR DESCRIPTION
**Pull request by gamesh411**
Make ODR violations warnings when using ASTImporter, but leave them errors when using Sema.
Please have a look at it before I send the patch to Phabricator.

### CODE REVIEW

File: include/clang/AST/ASTStructuralEquivalence.h [hunk 1]
```diff
@@ -70,9 +73,11 @@ struct StructuralEquivalenceContext {
   StructuralEquivalenceContext(ASTContext &FromCtx, ASTContext &ToCtx,
                                StructuralEquivalenceKind EqKind,
                                bool StrictTypeSpelling = false,
-                               bool Complain = true)
+                               bool Complain = true,
+                               bool ErrorOnTagTypeMismatch = false)
       : FromCtx(FromCtx), ToCtx(ToCtx), EqKind(EqKind),
-        StrictTypeSpelling(StrictTypeSpelling), Complain(Complain) {}
+        StrictTypeSpelling(StrictTypeSpelling),
+        ErrorOnTagTypeMismatch(ErrorOnTagTypeMismatch), Complain(Complain) {}
```
Comment (balazske) [inline include/clang/AST/ASTStructuralEquivalence.h]: The `ErrorOnTagTypeMismatch` belongs to the end of the initializer list.

File: include/clang/Basic/DiagnosticASTKinds.td [hunk 2]
```diff
@@ -228,70 +242,121 @@ def note_odr_number_of_bases : Note<
...
 def warn_odr_non_type_parameter_type_inconsistent : Warning<
   "non-type template parameter declared with incompatible types in different "
-  "translation units (%0 vs. %1)">, InGroup<OneDefinitionRule>;
+  "translation units (%0 vs. %1)">,
+  InGroup<ODR>;
 def err_unsupported_ast_node: Error<"cannot import unsupported AST node %0">;
+def warn_unsupported_ast_node: Warning<"cannot import unsupported AST node %0">,
+  InGroup<ODR>;
+def err_ast_importer_missing_decl_in_decl_context : Error<
+  "missing %0Decl in imported DeclContext <%1:%2:%3>">;
```
Comment (balazske) [inline include/clang/Basic/DiagnosticASTKinds.td]: Should this be an error?

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 3]
```diff
@@ -1305,7 +1334,10 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
 
   if (EC2 != EC2End) {
     if (Context.Complain) {
-      Context.Diag2(D2->getLocation(), diag::warn_odr_tag_type_inconsistent)
+      Context.Diag2(D2->getLocation(),
+                    Context.ErrorOnTagTypeMismatch
+                        ? diag::err_odr_tag_type_inconsistent
+                        : diag::warn_odr_tag_type_inconsistent)
```
Comment (balazske) [inline lib/AST/ASTStructuralEquivalence.cpp]: Maybe use a small function or lambda for these repeated statements (get constant for error or warning). (But a separate is needed for every kind of error-warning.)

File: test/ASTMerge/property/test.m [hunk 4]
```diff
@@ -10,4 +10,4 @@
 // CHECK: property2.m:29:21: note: property is synthesized to ivar 'ivar2' here
 // CHECK: property1.m:29:10: warning: property 'Prop3' is implemented with @dynamic in one translation but @synthesize in another translation unit
 // CHECK: property2.m:31:13: note: property 'Prop3' is implemented with @synthesize here
-// CHECK: 4 warnings generated.
+// CHECK: 4 warnings  generated.
```
Comment (balazske) [inline test/ASTMerge/property/test.m]: Why is this change needed?

File: include/clang/AST/ASTStructuralEquivalence.h (hunk 1 above)
Comment (gamesh411) [reply include/clang/AST/ASTStructuralEquivalence.h]: Done

File: include/clang/Basic/DiagnosticASTKinds.td (hunk 2 above)
Comment (gamesh411) [reply include/clang/Basic/DiagnosticASTKinds.td]: Returned these to their original state.

File: test/ASTMerge/property/test.m (hunk 4 above)
Comment (gamesh411) [reply test/ASTMerge/property/test.m]: Thanks, this was a typo.

File: lib/AST/ASTStructuralEquivalence.cpp (hunk 3 above)
Comment (gamesh411) [reply lib/AST/ASTStructuralEquivalence.cpp]: Implemented a quick lookup method.

**Comment by balazske** [comment]
The `ASTImporter` generates now only warnings for ODR mismatch, the structural equivalence may generate error or warning. My observation is only that there are warning-error pairs and code in `getApplicableDiagnostic` for every existing error type but many of these are not used in ASTStructuralEquivalence. These are used in `ASTImporter` but only with warning (many of the obj-c related things). It looks better to have uniform handling for every type (the current code) but now there are unused error codes. Probably later these should be used if the structural equivalence code is improved to handle the obj-c cases that currently produce warning in `ASTImporter`.

**Comment by gamesh411** [comment]
I agree. There are some diagnostics which we don't use currently, however we might later. I would like the code to work this way: if the ErrorOnTagTypeMismatch flag is set, then automatically get the error for a bug category (err_odr_tag_type_inconsistent in case the detected fault is due to inconsistent tag types), otherwise the warning (warn_odr_tag_type_inconsistent). I think associating errors with their corresponding warnings is what is missing from the diagnostic library, as they are now modeled as unrelated diagnostics.

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 5]
```diff
@@ -1177,15 +1187,15 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
...
-              Friend2End = D2CXX->friend_end();
+                                     Friend2End = D2CXX->friend_end();
       for (CXXRecordDecl::friend_iterator Friend1 = D1CXX->friend_begin(),
-                   Friend1End = D1CXX->friend_end();
+                                          Friend1End = D1CXX->friend_end();
            Friend1 != Friend1End; ++Friend1, ++Friend2) {
         if (Friend2 == Friend2End) {
           if (Context.Complain) {
             Context.Diag2(D2->getLocation(),
                           diag::warn_odr_tag_type_inconsistent)
```
Comment (martong) [inline lib/AST/ASTStructuralEquivalence.cpp]: Why don't we use `getApplicableDiagnostic()` here too?

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 6]
```diff
@@ -1194,8 +1204,9 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
 
         if (!IsStructurallyEquivalent(Context, *Friend1, *Friend2)) {
           if (Context.Complain) {
-            Context.Diag2(D2->getLocation(), diag::warn_odr_tag_type_inconsistent)
-              << Context.ToCtx.getTypeDeclType(D2CXX);
+            Context.Diag2(D2->getLocation(),
+                          diag::warn_odr_tag_type_inconsistent)
```
Comment (martong) [inline lib/AST/ASTStructuralEquivalence.cpp]: Missing use of `getApplicableDiagnostic()` ?

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 7]
```diff
@@ -1206,7 +1217,7 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
       if (Friend2 != Friend2End) {
         if (Context.Complain) {
           Context.Diag2(D2->getLocation(), diag::warn_odr_tag_type_inconsistent)
-                  << Context.ToCtx.getTypeDeclType(D2);
+              << Context.ToCtx.getTypeDeclType(D2);
```
Comment (martong) [inline lib/AST/ASTStructuralEquivalence.cpp:1226]: Here too.

**Review by martong — COMMENTED** [review]
Please try to avoid formatting unrelated lines, though this time it is okayish, because there is not so much of those changes.

File: lib/AST/ASTStructuralEquivalence.cpp (hunk 6 above)
Comment (gamesh411) [reply lib/AST/ASTStructuralEquivalence.cpp]: These are good catches! I was grepping for Context.*, and these slipped under the radar. With the next commit, all references to warnings were purged, which were not inside the getApplicableDiagnostic function.

**Review by martong — APPROVED** [review]
LG!

**Review by balazske — APPROVED** [review]
